Update to Nock 4K

cmarcelo opened the pull request.

lukechampine commented:
Thanks for doing this!

I was surprised to discover recently that Nock 4K is somewhat controversial...or at least, ~fyr doesn't like it. So I'm wondering if maybe Nock 4K and Nock 5K can live alongside each other. Maybe in separate subfolders, or on different branches. What do you think?

cmarcelo commented:
> I was surprised to discover recently that Nock 4K is somewhat controversial...or at least, ~fyr doesn't like it. So I'm wondering if maybe Nock 4K and Nock 5K can live alongside each other. Maybe in separate subfolders, or on different branches. What do you think?

_If_ we do want to keep both, there is significant reuse that I'm thinking about simply keeping them in the same package as two separate Nock functions: `Nock4()` and `Nock5()`. Perhaps make an alias for one of them as `Nock()`. What do you think?

lukechampine commented:
That's true. Ok, let's make `Nock()` Nock 4K (as it currently is in this PR) and also add a `Nock5K()` function.

cmarcelo committed:
Use a switch and straight code for tar

Instead of creating a map and a set of closures, just switch and run
the code for each instruction.  Add an explicit panic message for
unknown instructions.

cmarcelo committed:
Don't be permissive about Nock input

Neither an atom or an cell with an atom tail are valid inputs.  This
should panic or if/when there's a failure report mechanism, a failure.

cmarcelo committed:
Make the Nock version explicit in tar

Also expose an explicit Nock5 function.  This is a preparation for
adding support for the newer Nock spec while maintaining the old one
around.

cmarcelo committed:
Add Nock 4 support

Note Nock uses "Kelvin versioning", so the smallest version is the
latest.  Change the default to be Nock 4.  The old one is still
available as Nock5 function.

Tests were split into a common set, and specific tests for each
version.

cmarcelo committed:
Add short title for each instruction

These are not official titles but help understanding.  Based on
https://blog.timlucmiptev.space/ Nock tutorials.

cmarcelo commented:
There's Nock4 and Nock5, and Nock just calls Nock4. Also made other changes along the way.

@lukechampine could you take another look?

lukechampine commented on nock.go:
```diff
@@ -66,56 +66,175 @@ func fas(i int, n Noun) Noun {
 	}
 	inst, arg := form.Head(), form.Tail()
-	return map[int]func() Noun{
+	switch inst.Num() {
```
I guess the map literal approach was a little unsafe... :P

lukechampine commented:
ok, LGTM! I'll send your planet over shortly :)
